### DISCUSSION TITLE
Basics of the improved Sendbird ChatSDK initialization.

### DISCUSSION THREAD

**Discussion by tezpark** [sendbird/sendbird-chat-sdk-ios#34] [Commons]
![Languages](https://img.shields.io/badge/Swift-FA7343?logo=swift&logoColor=white) [![Version](https://img.shields.io/badge/Version-4.15.1-blue.svg)](https://github.com/sendbird/sendbird-chat-sdk-ios/releases/tag/v4.15.1)

### Basics of the improved Sendbird ChatSDK initialization.

Now, initialize the Chat SDK in the app to allow the Chat SDK to respond to changes according to the connection status in iOS client apps. Initialization requires your Sendbird application's `APP_ID`, which can be found on [Sendbird Dashboard](https://dashboard.sendbird.com).

Before calling the [`initialize(params:migrationStartHandler:completionHandler:)`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/SendbirdChat.html#/c:@M@SendbirdChatSDK@objc(cs)SendbirdChat(cm)initializeWithParams:migrationStartHandler:completionHandler:) method, create an [`InitParams`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/InitParams.html) object with your Sendbird application ID. In the params, you can also determine whether to enable [local caching](https://sendbird.com/docs/chat/sdk/v4/ios/local-caching/overview-local-caching) or not and whether to make initialization synchronously.

<div component="AdvancedCode" title="Swift">

```swift,chat/v4/ios/getting-started/send-first-message-176
let initParams = InitParams(
    applicationId: APP_ID,
    isLocalCachingEnabled: true,
    logLevel: .info,
    needsSynchronous: false
)
SendbirdChat.initialize(
    params: params,
    migrationStartHandler: {
        // Migration starts.    
    },
    completionHandler: { error in
        // Migration completed.    
    }
)
```

</div>

The `completionHandler` method gets the initialization status through different event handlers. The `migrationStartHandler` method is called when there's an upgrade in the local database. Meanwhile, the `completionHandler` method informs the client app whether the initialization is complete.

If the initialization fails when you set [`isLocalCachingEnabled`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/InitParams.html#/c:@M@SendbirdChatSDK@objc(cs)SBDInitParams(py)isLocalCachingEnabled) to `true`, the SDK will operate normally and change the value of [`isLocalCachingEnabled`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/InitParams.html#/c:@M@SendbirdChatSDK@objc(cs)SBDInitParams(py)isLocalCachingEnabled) to `false`. If you still wish to use local caching, clear the database using [`clearCachedData(completionHandler:)`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/SendbirdChat.html#/c:@M@SendbirdChatSDK@objc(cs)SendbirdChat(cm)clearCachedDataWithCompletionHandler:) and try the initialization again with [`isLocalCachingEnabled`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/InitParams.html#/c:@M@SendbirdChatSDK@objc(cs)SBDInitParams(py)isLocalCachingEnabled) set to `true`.

If you are using local caching but don't need to use Sendbird's features immediately after launching your app, you can make the migration process of local caching asynchronous by setting the `needsSynchronous` parameter of the [`InitParams`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/InitParams.html) object to `false`. This will reduce the initialization time when the app is launched to minimize runtime. (If you don't use local caching, there's no difference in app launching time because there's no migration process).

>__Note__: The [`initialize(params:migrationStartHandler:completionHandler:)`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/SendbirdChat.html#/c:@M@SendbirdChatSDK@objc(cs)SendbirdChat(cm)initializeWithParams:migrationStartHandler:completionHandler:) method of a [`SendbirdChat`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/SendbirdChat.html) instance must be called across a client app at least once. We recommend that you initialize Sendbird Chat SDK through the [`application:didFinishLaunchingWithOptions:`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) method of the `AppDelegate` instance.

Also, if you have a situation where a SendbirdChat function is called asynchronously from another class or method, independent of SendbirdChat's initialization, use the SendbirdChat class's `executeOrWaitForInitialization(executeHandler:)` feature. 
You can set it to execute immediately if it's initialized, or wait for initialization if it's not.

<div component="AdvancedCode" title="Swift">

```swift,chat/v4/ios/getting-started/send-first-message-207
SendbirdChat.executeOrWaitForInitialization {
    // Implement logic to perform once initialization is complete
}
```

>__Note__: The `executeOrWaitForInitialization(executeHandler:)` method of a [`SendbirdChat`](https://sendbird.com/docs/chat/sdk/v4/ios/ref/Classes/SendbirdChat.html) instance must be set after calling Chat initialization method.

</div>

#### Sequence diagram

Below is a sequence diagram to help you understand.

##### Synchronous

![sd Chat - initialize_ Sync (Tobe)](https://github.com/sendbird/sendbird-chat-sdk-ios/assets/389004/8d55d957-2904-4d72-9825-f34f6f598d0f)

##### Asynchronous

![sd Chat - initialize_ Async (Tobe)](https://github.com/sendbird/sendbird-chat-sdk-ios/assets/389004/778d199e-fce1-4628-9eb5-bb219755d409)

**Comment by fizix100** [comment]
The recent versions of SendbirdChatSDK iOS (4.27.1~4.28.0) have a serious issue where memory and CPU usage remain consistently high. Memory consumption keeps increasing until a memory usage warning is triggered.
<img width="3852" height="8334" alt="image" src="https://github.com/user-attachments/assets/74d54bf0-43ac-45f5-a53d-2d5314164ff6" />
<img width="2462" height="1620" alt="WX20250722-112217@2x" src="https://github.com/user-attachments/assets/ec9eaffa-70ab-4034-92dc-cda3bfe65464" />
